### PR TITLE
Adding RayHistory to DagMC

### PR DESCRIPTION
**Pull request by pshriwise**
This PR includes a the use of a ray history for DagMC runs.  When a history is passed to `DagMC::ray_fire`, all triangles in that history are masked from the ray query and allow the particle to continue to the next intersection. This construct helps handle some numerical issues with sharply angled surfaces like the ones shown below.

angled_surfs

If a particle starts in volume **B** intersects right on an edge shared by these two surfaces (whose normals are shown using the color of their respective surfaces below). The desired behavior here is a zero-distance path through volume **A** and  continuation into volume **B**. 

step1

At some point during transport, the particle ray intersects a facet on the blue surface - specifically the blue surface because DagMC searches only for exiting intersections for the volume the particle is currently in. An exiting intersection meaning an intersection with a triangle whose normal is aligned with the ray direction.

step2

The particle then moves into volume **A** via the a `DagMC::next_vol` call. At this point, triangle normals are flipped due to the new containment of the particle. A zero-distance intersection with the *red* surface is then found and we move back into volume **B**. So far so good, but because these facets aren't tracked in a history another zero-distance intersection with the blue surface is found. The particle moves into volume **A**, and the process starts all over again.

With a ray history in place and passed to `ray_fire`, both of the facets adjacent to that edge will be masked and the particle can continue into volume **B** as expected.

### Resetting Ray Histories 

These histories avoid problems like the one above, but can also cause problems if they aren't managed properly. One example is the scenario in which a particle scatters through a very small angle and strikes the same facet it did previously. If that facet is masked out, then no intersection is found and the particle is falsely claimed as lost. Therefore the history is reset any time the direction changes. Additionally, histories are reset when moving into a new volume, so that a backscattering particle exit that volume through the same facet it entered through. This is the equivalent of resetting a history when a particle is not on a surface. Finally, when a new particle is created, the history should be reset so that a coincidental position and direction intersecting with a triangle in the history doesn't result in an immediate lost particle.

  - reset when particle direction changes
  - reset when moving off a surface in a new volume
  - reset when a new particle is created

A less obvious case is reflecting surfaces in void regions, if a particle moves through such a void region it will go from surface to surface of the volume. With the history reset rules in the previous paragraph applied, there will then be two triangles in the ray history, the entrance triangle and the exit triangle on the reflecting surface. If, by chance, the particle reflects back toward the facet it came from this triangle will be masked out &rarr; no intersection &rarr; false lost particle. Conceptually, this is the same as the new volume case, but we'll want to reflect the history as 

  - reset when reflecting a particle

This PR also:

  - places DagMC tests in their own subdirectory
  - adds a reflecting boundary condition DagMC regression test

### CODE REVIEW

**Comment by pshriwise** [comment]
I also meant to note that threaded implementations of DagMC using a ray history aren't common. There is one other that I know of in Frensie. It's a valuable feature to have as more complicated DagMC models can take up lot of memory (up to ~10 GB) and limit the number of processes per node in HPC environments.

File: include/openmc/dagmc.h [hunk 1]
```diff
@@ -9,11 +9,15 @@ extern "C" const bool dagmc_enabled;
...
-#include "openmc/cell.h"
-#include "openmc/surface.h"
+#include "openmc/xml_interface.h"
+#include "openmc/position.h"
 
 namespace openmc {
 
+extern moab::DagMC::RayHistory history;
+extern Direction last_dir;
+#pragma omp threadprivate(history, last_dir)
```
Comment (paulromano) [inline include/openmc/dagmc.h]: Does this work on the intel compiler?

File: include/openmc/dagmc.h [hunk 2]
```diff
@@ -9,11 +9,15 @@ extern "C" const bool dagmc_enabled;
...
 #include "DagMC.hpp"
-#include "openmc/cell.h"
-#include "openmc/surface.h"
+#include "openmc/xml_interface.h"
+#include "openmc/position.h"
 
 namespace openmc {
 
+extern moab::DagMC::RayHistory history;
+extern Direction last_dir;
```
Comment (paulromano) [inline include/openmc/dagmc.h]: Global variables should be in a [sub-namespace](http://docs.openmc.org/en/latest/devguide/styleguide.html#naming).

File: src/particle.cpp [hunk 3]
```diff
@@ -372,6 +372,12 @@ Particle::transport()
     }
   }
 
+  #ifdef DAGMC
+  if(settings::dagmc) {
+    history.reset();
+  }
```
Comment (paulromano) [inline src/particle.cpp]: `if (settings::dagmc) history.reset();`

File: tests/regression_tests/dagmc/uwuw/test.py [hunk 4]
```diff
@@ -0,0 +1,41 @@
+import openmc
+import openmc.capi
+from openmc.stats import Box
+from openmc.material import Materials
```
Comment (paulromano) [inline tests/regression_tests/dagmc/uwuw/test.py]: `openmc.material` is not really meant to be a user-visible module. Use `from openmc import Materials` or just call it out below as `openmc.Materials`

File: include/openmc/dagmc.h (hunk 1 above)
Comment (pshriwise) [reply include/openmc/dagmc.h]: It does. I think because they aren't templated types?

Comment (paulromano) [reply include/openmc/dagmc.h]: right, makes sense

File: include/openmc/dagmc.h (hunk 2 above)
Comment (pshriwise) [reply include/openmc/dagmc.h]: `model` or `simulation` would you say?

File: tests/regression_tests/dagmc/uwuw/test.py (hunk 4 above)
Comment (pshriwise) [reply tests/regression_tests/dagmc/uwuw/test.py]: On top of that, I don't use `Materials` in those tests anymore. Removing. That's good to know for the future, though!

File: include/openmc/dagmc.h (hunk 2 above)
Comment (pshriwise) [reply include/openmc/dagmc.h]: I went with `simulation` as that's when they are used, let me know if I'm off on that.

Comment (paulromano) [reply include/openmc/dagmc.h]: Yup, `simulation` makes the most sense

**Comment by paulromano** [comment]
@makeclean Any comments on this PR?

File: include/openmc/cell.h [hunk 5]
```diff
@@ -11,7 +11,7 @@
 #include "hdf5.h"
 #include "pugixml.hpp"
 #ifdef DAGMC
-#include "DagMC.hpp"
+#include "dagmc.h"
```
Comment (makeclean) [inline include/openmc/cell.h]: is this a bespoke part of openmc?

Comment (pshriwise) [reply include/openmc/cell.h]: The `DagMC.hpp` include is, but `dagmc.h` isn't. This `#ifdef` can be removed then I think. 
Thanks!
